gammapy/detect/test_statistics.py

```diff
@@ -72,30 +93,130 @@ def f_cash(x, counts, background, model):
     ----------
     x : float
         Model amplitude.
-    counts : `~numpy.ndarray`
+    counts : array
         Count map slice, where model is defined.
-    background : `~numpy.ndarray`
+    background : array
         Background map slice, where model is defined.
-    model : `~numpy.ndarray`
+    model : array
         Source template (multiplied with exposure).
     """
     with np.errstate(invalid='ignore', divide='ignore'):
         return cash(counts, background + x * FLUX_FACTOR * model).sum()
 
 
-def compute_ts_map(counts, background, exposure, kernel, flux=None,
-                   method='root', optimizer='Brent', parallel=True,
+def compute_ts_map_multiscale(maps, psf_parameters, scales=[0], downsample='auto',
+                              residual=False, morphology='Gaussian2D', width=0.3,
+                              *args, **kwargs):
+    """
+    Compute multiscale TS maps using compute_ts_map.
+
+    High level TS map computation using a multi gauss PSF kernel and assuming
+    a given source morphology. To optimize the performance the input data
+    can be sampled down when computing TS maps on larger scales.
+
+    Parameters
+    ----------
+    maps : `astropy.io.fits.HDUList`
+        HDU list containing the data. The list must contain the following HDU
+        extensions:
+            * 'On' -- Counts image
+            * 'Background' -- Background image
+            * 'Diffuse' -- Diffuse model image
+            * 'ExpGammaMap' -- Exposure image
+    psf_parameters : dict
+        Dict defining the multi gauss PSF parameters.
+        See `~gammapy.irf.multi_gauss_psf` for details.
+    scales : list ([0])
+        List of scales to use for TS map computation.
+    downsample : int (auto)
+        Down sampling factor. Can be set to 'auto' if the down sampling
+        factor should be chosen automatically.
+    residual : bool (False)
+        Compute a TS residual map.
+    morphology : str ('Gaussian2D')
+        Source morphology assumption. Either 'Gaussian2D' or 'Shell2D'.
+
+    Returns
+    -------
+    multiscale_result : list
+        List of `TSMapResult` objects.
+    """
+    BINSZ = abs(maps[0].header['CDELT1'])
+    shape = maps[0].data.shape
+    multiscale_result = []
+
+    for scale, factor in zip(scales, downsample):
+        logging.info('Computing (residual) TS map for scale {0} deg'.format(scale))
+
+        # Sample down and require that scale parameters is at least 5 pix
```

Comment:
This function is quite long ... ideally every function should fit on a screen.

Can you split it up, e.g. extract this part into a `_downsample` helper function?